### PR TITLE
Draw line of a route name to the end of row console on `rails routes --expanded`

### PR DESCRIPTION
**Pull request by bogdanvlviv**
In order to get width of console use `IO::console_size`,
See https://ruby-doc.org/stdlib-2.4.1/libdoc/io/console/rdoc/IO.html#method-c-console_size

Related to #32130

### CODE REVIEW

**Comment by kaspth** [comment]
Sure we can make this dynamic, but what difference does it really make if the dashes go to the end of a window?

I don't care much for the extra dynamicness/variability it adds. Especially in the tests.

I'll let other reopen if they're swayed differently. Thanks!

**Comment by bogdanvlviv** [comment]
When a window has a width less than 73, `rails routes --expanded` will output lines like
```bash
$ rails routes --expanded
--[ Route 1 ]------------------------------------------------------------
-------
(snip)
--[ Route 42 ]-----------------------------------------------------------
--------
(snip)
--[ Route 333 ]----------------------------------------------------------
---------
(snip)
```
This  PR make the output of `rails routes --expanded` neater in this case.
```bash
$ rails routes --expanded
--[ Route 1 ]------------------------------------------------------------
(snip)
--[ Route 42 ]-----------------------------------------------------------
(snip)
--[ Route 333 ]----------------------------------------------------------
(snip)
```

**Comment by matthewd** [comment]
> When a window has a width less than 73

That sounds pretty compelling to me; I hate when that happens.

Let's find a way of overriding the width detection in the test, though.

**Comment by bogdanvlviv** [comment]
@matthewd We can use `IO.console.winsize=` in order to override the width of console in tests?
```diff
diff --git a/actionpack/test/dispatch/routing/inspector_test.rb b/actionpack/test/dispatch/routing/inspector_test.rb
index e970ef0..f352e7a 100644
--- a/actionpack/test/dispatch/routing/inspector_test.rb
+++ b/actionpack/test/dispatch/routing/inspector_test.rb
@@ -323,6 +323,9 @@ def test_routes_can_be_filtered
       end
 
       def test_routes_when_expanded
+        previous_console_winsize = IO.console.winsize
+        IO.console.winsize = [0, 0]
+
         engine = Class.new(Rails::Engine) do
           def self.inspect
             "Blog::Engine"
@@ -338,31 +341,30 @@ def self.inspect
           mount engine => "/blog", :as => "blog"
         end
 
-        line_width = IO.console_size.second - 13
-        line_width = 0 if line_width.negative?
-
-        assert_equal ["--[ Route 1 ]#{'-' * line_width}",
+        assert_equal ["--[ Route 1 ]",
                       "Prefix            | custom_assets",
                       "Verb              | GET",
                       "URI               | /custom/assets(.:format)",
                       "Controller#Action | custom_assets#show",
-                      "--[ Route 2 ]#{'-' * line_width}",
+                      "--[ Route 2 ]",
                       "Prefix            | custom_furnitures",
                       "Verb              | GET",
                       "URI               | /custom/furnitures(.:format)",
                       "Controller#Action | custom_furnitures#show",
-                      "--[ Route 3 ]#{'-' * line_width}",
+                      "--[ Route 3 ]",
                       "Prefix            | blog",
                       "Verb              | ",
                       "URI               | /blog",
                       "Controller#Action | Blog::Engine",
                       "",
                       "[ Routes for Blog::Engine ]",
-                      "--[ Route 1 ]#{'-' * line_width}",
+                      "--[ Route 1 ]",
                       "Prefix            | cart",
                       "Verb              | GET",
                       "URI               | /cart(.:format)",
                       "Controller#Action | cart#show"], output
+      ensure
+        IO.console.winsize = previous_console_winsize
       end
 
 
diff --git a/railties/test/commands/routes_test.rb b/railties/test/commands/routes_test.rb
index d8c9b98..3bf7b23 100644
--- a/railties/test/commands/routes_test.rb
+++ b/railties/test/commands/routes_test.rb
@@ -120,54 +120,58 @@ class Rails::Command::RoutesTest < ActiveSupport::TestCase
   end
 
   test "rails routes with expanded option" do
-    app_file "config/routes.rb", <<-RUBY
-        Rails.application.routes.draw do
-          get '/cart', to: 'cart#show'
-        end
-    RUBY
-
-    output = rails("routes", "--expanded")
+    begin
+      previous_console_winsize = IO.console.winsize
+      IO.console.winsize = [0, 0]
 
-    line_width = IO.console_size.second - 13
-    line_width = 0 if line_width.negative?
-
-    assert_equal <<~MESSAGE, output
-    --[ Route 1 ]#{'-' * line_width}
-    Prefix            | cart
-    Verb              | GET
-    URI               | /cart(.:format)
-    Controller#Action | cart#show
-    --[ Route 2 ]#{'-' * line_width}
-    Prefix            | rails_service_blob
-    Verb              | GET
-    URI               | /rails/active_storage/blobs/:signed_id/*filename(.:format)
-    Controller#Action | active_storage/blobs#show
-    --[ Route 3 ]#{'-' * line_width}
-    Prefix            | rails_blob_variation
-    Verb              | GET
-    URI               | /rails/active_storage/variants/:signed_blob_id/:variation_key/*filename(.:format)
-    Controller#Action | active_storage/variants#show
-    --[ Route 4 ]#{'-' * line_width}
-    Prefix            | rails_blob_preview
-    Verb              | GET
-    URI               | /rails/active_storage/previews/:signed_blob_id/:variation_key/*filename(.:format)
-    Controller#Action | active_storage/previews#show
-    --[ Route 5 ]#{'-' * line_width}
-    Prefix            | rails_disk_service
-    Verb              | GET
-    URI               | /rails/active_storage/disk/:encoded_key/*filename(.:format)
-    Controller#Action | active_storage/disk#show
-    --[ Route 6 ]#{'-' * line_width}
-    Prefix            | update_rails_disk_service
-    Verb              | PUT
-    URI               | /rails/active_storage/disk/:encoded_token(.:format)
-    Controller#Action | active_storage/disk#update
-    --[ Route 7 ]#{'-' * line_width}
-    Prefix            | rails_direct_uploads
-    Verb              | POST
-    URI               | /rails/active_storage/direct_uploads(.:format)
-    Controller#Action | active_storage/direct_uploads#create
-    MESSAGE
+      app_file "config/routes.rb", <<-RUBY
+          Rails.application.routes.draw do
+            get '/cart', to: 'cart#show'
+          end
+      RUBY
+
+      output = rails("routes", "--expanded")
+
+      assert_equal <<~MESSAGE, output
+      --[ Route 1 ]
+      Prefix            | cart
+      Verb              | GET
+      URI               | /cart(.:format)
+      Controller#Action | cart#show
+      --[ Route 2 ]
+      Prefix            | rails_service_blob
+      Verb              | GET
+      URI               | /rails/active_storage/blobs/:signed_id/*filename(.:format)
+      Controller#Action | active_storage/blobs#show
+      --[ Route 3 ]
+      Prefix            | rails_blob_variation
+      Verb              | GET
+      URI               | /rails/active_storage/variants/:signed_blob_id/:variation_key/*filename(.:format)
+      Controller#Action | active_storage/variants#show
+      --[ Route 4 ]
+      Prefix            | rails_blob_preview
+      Verb              | GET
+      URI               | /rails/active_storage/previews/:signed_blob_id/:variation_key/*filename(.:format)
+      Controller#Action | active_storage/previews#show
+      --[ Route 5 ]
+      Prefix            | rails_disk_service
+      Verb              | GET
+      URI               | /rails/active_storage/disk/:encoded_key/*filename(.:format)
+      Controller#Action | active_storage/disk#show
+      --[ Route 6 ]
+      Prefix            | update_rails_disk_service
+      Verb              | PUT
+      URI               | /rails/active_storage/disk/:encoded_token(.:format)
+      Controller#Action | active_storage/disk#update
+      --[ Route 7 ]
+      Prefix            | rails_direct_uploads
+      Verb              | POST
+      URI               | /rails/active_storage/direct_uploads(.:format)
+      Controller#Action | active_storage/direct_uploads#create
+      MESSAGE
+    ensure
+      IO.console.winsize = previous_console_winsize
+    end
   end
 
   private

```

**Comment by kaspth** [comment]
@bogdanvlviv `winsize` looks good 👍 — maybe we should set it to 10 so we can see some lines after `--[ Route 1 ]` being printed.

**Comment by kaspth** [comment]
By the way, how much do the `[ Route x ]` help versus just dashes all the way? Nobody would refer to a route by index — especially because the index changes.

**Comment by bogdanvlviv** [comment]
@kaspth We should set `winsize` to  `> 13` in order to see some lines, because `--[ Route 1 ]` takes 13

> By the way, how much do the [ Route x ] help versus just dashes all the way? Nobody would refer to a route by index — especially because the index changes.

It was designed in almost the same way as `psql` does - https://github.com/rails/rails/pull/32130#issue-171781741, but we definitely can change this to dashes only.

**Comment by kaspth** [comment]
👍, set it to 23 or more then, just so we can see the extra dashes get printed 😊

> It was designed in almost the same way as psql does - #32130 (comment), but we definitely can change this to dashes only.

I guess it's useful when a search returns > 20 results and it gives you a better sense of where you are in the list. E.g. scroll to route ~6, then route ~16, realize you want to go back to up to around where you saw that other ~6 route. Supposedly that would be easier if there's a number. Personally, I'm ¯\_(ツ)_/¯ on them.

File: actionpack/lib/action_dispatch/routing/inspector.rb [hunk 1]
```diff
@@ -223,9 +223,16 @@ def no_routes(routes)
         private
 
           def draw_expanded_section(routes)
+            require "io/console/size"
+            console_width = IO.console_size.second
+
             routes.map.each_with_index do |r, i|
+              section_name = "--[ Route #{i + 1} ]"
+              line_width = console_width - section_name.size
+              line_width = 0 if line_width.negative?
```
Comment (kaspth) [inline actionpack/lib/action_dispatch/routing/inspector.rb]: This seems like it's working at a separate level of abstraction. How about extracting it to a `route_header` method: `route_header(index: i + 1)`?

```ruby
def route_header(index:)
  @console_width ||= begin
    require "io/console/size"
    IO.console_size.second
  end

  header_prefix     = "--[ Route #{index} ]"
  dash_remainder = [ console_width - header_prefix.size, 0 ].max
  
  "#{header_prefix}#{"-" * dash_remainder}"
end
```

Comment (bogdanvlviv) [reply actionpack/lib/action_dispatch/routing/inspector.rb]: Looks like we can omit ivar.
```diff
diff --git a/actionpack/lib/action_dispatch/routing/inspector.rb b/actionpack/lib/action_dispatch/routing/inspector.rb
index 19dfb44..40fd007 100644
--- a/actionpack/lib/action_dispatch/routing/inspector.rb
+++ b/actionpack/lib/action_dispatch/routing/inspector.rb
@@ -102,6 +102,7 @@ def filter_routes(filter)
         end

         def collect_routes(routes)
+          Array.new(1000,
           routes.collect do |route|
             RouteWrapper.new(route)
           end.reject(&:internal?).collect do |route|
@@ -112,6 +113,7 @@ def collect_routes(routes)
               path: route.path,
               reqs: route.reqs }
           end
+          ).flatten
         end

         def collect_engine_routes(route)
```
```bash
blog$ time rails routes --expanded
(snip)
--[ Route 13000 ]-----------------------------------------------------------------------------------------------------------
Prefix            | rails_direct_uploads
Verb              | POST
URI               | /rails/active_storage/direct_uploads(.:format)
Controller#Action | active_storage/direct_uploads#create

real    0m1.483s
user    0m1.037s
sys     0m0.337s
```

<hr/>

```diff
diff --git a/actionpack/lib/action_dispatch/routing/inspector.rb b/actionpack/lib/action_dispatch/routing/inspector.rb
index 19dfb44..4536a52 100644
--- a/actionpack/lib/action_dispatch/routing/inspector.rb
+++ b/actionpack/lib/action_dispatch/routing/inspector.rb
@@ -102,6 +102,7 @@ def filter_routes(filter)
         end

         def collect_routes(routes)
+          Array.new(1000,
           routes.collect do |route|
             RouteWrapper.new(route)
           end.reject(&:internal?).collect do |route|
@@ -112,6 +113,7 @@ def collect_routes(routes)
               path: route.path,
               reqs: route.reqs }
           end
+          ).flatten
         end

         def collect_engine_routes(route)
@@ -189,6 +191,7 @@ def widths(routes)
       class Expanded
         def initialize
           @buffer = []
+          @console_width = IO.console_size.second
         end

         def result
@@ -236,9 +239,8 @@ def draw_expanded_section(routes)
           end

           def route_header(index:)
-            console_width = IO.console_size.second
             header_prefix = "--[ Route #{index} ]"
-            dash_remainder = [console_width - header_prefix.size, 0].max
+            dash_remainder = [@console_width - header_prefix.size, 0].max

             "#{header_prefix}#{'-' * dash_remainder}"
           end
```

```bash
blog$ time rails routes --expanded
(snip)
--[ Route 13000 ]-----------------------------------------------------------------------------------------------------------
Prefix            | rails_direct_uploads
Verb              | POST
URI               | /rails/active_storage/direct_uploads(.:format)
Controller#Action | active_storage/direct_uploads#create

real    0m1.485s
user    0m1.107s
sys     0m0.269s

```

**Comment by benoittgt** [comment]
Thanks for the PR !

**Pull request merged**